En este archivo puedes escribir lo que estimes conveniente. Te recomendamos detallar tu solución y todas las suposiciones que estás considerando. Aquí puedes ejecutar las funciones que definiste en los otros archivos de la carpeta src, medir el tiempo, memoria, etc.

In [2]:
file_path = "farmers-protest-tweets-2021-2-4.json"

El objetivo de este algoritmo es identificar las top 10 fechas con más tweets y el usuario más activo en esas fechas. Para ello, se creo un algoritmo para el procesado de los twwts al usar las herramientas sugeridas se realizó un análisis de uso de memoria y tiempo de ejecución.

El inicio de memoria fue de 55.9 MB y se estabilizó en 61.6 MB, lo que sugiere un aumento moderado en el uso de memoria a medida que se procesan los datos. La línea que procesa los tweets (json.loads(line)) se ejecutó 117,408 veces, indicando un volumen considerable de datos. El tiempo total de ejecución fue de 37.033 segundos, lo que sugiere la presencia de un cuello de botella.

Para optimizar el rendimiento del algoritmo, propongo las siguientes mejoras:

Leer por bloques: Esto puede ayudar a reducir el uso de memoria al evitar cargar todo el archivo en memoria.
Usar Counter en lugar de defaultdict: Esto puede simplificar el código y mejorar el rendimiento.
Filtrar por fechas y usuarios: Dado que solo necesitamos las fechas más activas, contar las fechas primero y luego procesar esas fechas para obtener los usuarios más activos puede ser más eficiente.
En resumen, estas optimizaciones pueden resultar en un uso de memoria más eficiente y una mejora en el tiempo de ejecución.

In [4]:
from datetime import datetime
from q1_time import q1_time
from q1_memory import q1_memory

# Probar la función de tiempo
print("Probando q1_time...")
top_dates_time = q1_time(file_path)
print("Top 10 fechas con más tweets y el usuario más activo:")
for date, user in top_dates_time:
    print(f"Fecha: {date}, Usuario: {user}")

# Probar la función de memoria
print("\nProbando q1_memory...")
top_dates_memory = q1_memory(file_path)
print("Top 10 fechas con más tweets y el usuario más activo (medido en memoria):")
for date, user in top_dates_memory:
    print(f"Fecha: {date}, Usuario: {user}")


Probando q1_time...
Tiempo de ejecución: 4.20 segundos
Top 10 fechas con más tweets y el usuario más activo:
Fecha: 2021-02-19, Usuario: Preetm91
Fecha: 2021-02-18, Usuario: neetuanjle_nitu
Fecha: 2021-02-17, Usuario: RaaJVinderkaur
Fecha: 2021-02-13, Usuario: MaanDee08215437
Fecha: 2021-02-12, Usuario: RanbirS00614606
Fecha: 2021-02-21, Usuario: Surrypuria
Fecha: 2021-02-23, Usuario: Surrypuria
Fecha: 2021-02-15, Usuario: jot__b
Fecha: 2021-02-16, Usuario: jot__b
Fecha: 2021-02-14, Usuario: rebelpacifist

Probando q1_memory...
Filename: c:\Users\011101011110\Desktop\prueba\src\q1_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     7     78.6 MiB     78.6 MiB           1   @profile
     8                                         def q1_memory(file_path: str) -> List[Tuple[datetime.date, str]]:
     9     79.3 MiB     -5.0 MiB          27       date_user_count = defaultdict(lambda: defaultdict(int))
    10                                         
    11     78.6

Dando un resultado de 4.20 seg y un aumentod e memoria de 78.6 MiB a 79.6 MiB doy por terminado el ejericicio numero 1 habiendo sacado una mejora considerable y un tiempo standard de resultados esperados